sound/soc/sof/ipc4-topology.c

```diff
@@ -1322,10 +1324,28 @@ static int snd_sof_get_nhlt_endpoint_data(struct snd_sof_dev *sdev, struct snd_s
 	dev_dbg(sdev->dev, "dai index %d nhlt type %d direction %d\n",
 		dai_index, nhlt_type, dir);
 
+	if (dir == SNDRV_PCM_STREAM_PLAYBACK) {
+		nhlt_dirs[0] = NHLT_DIRECTION_RENDER;
+		nhlt_dirs[1] = NHLT_DIRECTION_RENDER_WITH_LOOPBACK;
+	} else {
+		nhlt_dirs[0] = NHLT_DIRECTION_CAPTURE;
+		nhlt_dirs[1] = NHLT_DIRECTION_FEEDBACK_FOR_RENDER;
+	}
+
 	/* find NHLT blob with matching params */
-	cfg = intel_nhlt_get_endpoint_blob(sdev->dev, ipc4_data->nhlt, dai_index, nhlt_type,
-					   bit_depth, bit_depth, channel_count, sample_rate,
-					   dir, 0);
+	for (i = 0; i < ARRAY_SIZE(nhlt_dirs); i++) {
+		for (j = 0; j < ARRAY_SIZE(nhlt_devices); j++) {
+			cfg = intel_nhlt_get_endpoint_blob(sdev->dev, ipc4_data->nhlt,
+							   dai_index, nhlt_type,
+							   bit_depth, bit_depth,
+							   channel_count, sample_rate,
+							   nhlt_dirs[i],
+							   nhlt_devices[j]);
+			if (cfg || (linktype == SOF_DAI_INTEL_DMIC))
+				goto cfg_found;
+		}
+	}
+cfg_found:
```

Comment:
the main issue I have is that this is not backwards compatible. You need a new topology to get this to work. Is this not too late already?

@ranj063 please chime in here.

Reply:
It's backward compatible. Because current implementation set the device type of all endpoints to 0 (BT), and use Render(0) and Capture(1) in direction field. So current topologies could work with the code.